lib/bap/bap.mli

```diff
@@ -3822,6 +3860,377 @@ module Std : sig
+        [id], if such exists.  *)
+    val next : ('a,'b) cls -> 'a term -> tid -> 'b term option
+
+    (** [next t p id] returns a term that is after a term with a given
+        [id], if such exists.  *)
+    val prev : ('a,'b) cls -> 'a term -> tid -> 'b term option
+
+    (** [after t ?rev p tid] returns all subterms in term [p] that occur
+        after a term with a given [tid]. if [rev] is [true] or omitted then
+        terms are returned in the evaluation order. Otherwise
+        they're reversed. If there is no term with a given [tid],
+        then an empty sequence is returned. *)
+    val after : ('a,'b) cls -> ?rev:bool -> 'a term -> tid -> 'b term seq
+
+    (** [before t ?rev p tid] returns all term that occurs before
+        defintion with given [tid] in blk. If there is no such
+        definition, then the sequence will be empty.  @param rev has the
+        same meaning as in {!after}.  *)
+    val before : ('a,'b) cls -> ?rev:bool -> 'a term -> tid -> 'b term seq
+
+    (** [append t ~after:this p c] returns the [p] term with [c]
+        appended after [this] term. If [after] is not specified,
+        then append [def] to the end of the parent term (if it makes
+        sense, otherwise it is just added).  If [this] doesn't
+        occur in the [p] term then do nothing. The term tid is
+        preserved.  *)
+    val append : ('a,'b) cls -> ?after:tid -> 'a term -> 'b term -> 'a term
+
+    (** [prepend t ~before:this p c] returns the [p] with [c] inserted
+        before [this] term. If [before] is left unspecified, then
+        insert the [c] at the beginning of the [p] if it is a sequence,
+        otherwise just insert. If [this] is specified but doesn't occur
+        in the [p] then [p] is returned as is.  In all cases, the
+        returned term has the same [tid] as [p]. *)
+    val prepend : ('a,'b) cls -> ?before:tid -> 'a term -> 'b term -> 'a term
+
+  end
+
+  (** Program.  *)
+  module Program : sig
+    (** Program is a collection of function terms. *)
+
+    type t = program term
+
+    (** [create ()] creates an empty program. *)
+    val create : unit -> t
+
+    (** [lookup t program id] is like {{!find}find} but performs deep
+        lookup in the whole [program] for a term with a given [id].
+        This function is memoized, so it has amortized O(1) complexity,
+        with a wostcase complexity of $O(N)$, where $N$ is the total
+        amount of terms in program.  *)
+    val lookup : (_,'b) cls -> t -> tid -> 'b term option
+
+    (** [parent t program id] is [Some p] iff [find t p id <> None]  *)
+    val parent : ('a,'b) cls -> t -> tid -> 'a term option
+
+    module Builder : sig
+      type t
+      val create : ?tid:tid  -> ?subs:int -> unit -> t
+      val add_sub : t -> sub term -> unit
+      val result : t -> program term
+    end
+
+    include Regular with type t := t
+  end
+
+  (** Function.  *)
+  module Sub : sig
```

Comment:
Hmm. I was thinking a bit differently of the abstraction where it wouldn't hide the lower level details, just simplify them.  ABI seems like it would be relevant to many analysis, e.g., an analysis on the IR may want to update how many arguments there really are based on its results, which in turn may be improved by considering the abi.  I'm not sure it's useful to hide this detail in the abstraction, rather than provide a summary.  We can wait and see, but that's my bet for now.